### PR TITLE
Implement Project.allowed_to, returning an ActiveRecord::Relation

### PR DESCRIPTION
**Pull request by myabc**
:warning: ~~**Includes commits from #3194. Please review, merge first.**~~

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://myabc-feature-20325-rails4-deprecated-dynamic-finders-allowed-t.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://myabc-feature-20325-rails4-deprecated-dynamic-finders-allowed-t.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by ulferts** [comment]
@myabc now that #3194 is through, can you shorten this PR, please?

**Comment by myabc** [comment]
referencing comment in fa011e178582051403ef60b63195b57d14d759b9  for posterity.

File: app/models/news.rb [hunk 1]
```diff
@@ -68,22 +67,14 @@ def visible?(user = User.current)
...
-    conditions = Project.allowed_to_condition(user, :view_news)
-
-    # TODO: remove the includes from here, it's required by Project.allowed_to_condition
-    # News has nothing to do with it
-    where(conditions).limit(limit).newest_first.includes(:author, :project)
-      .references(:users, :projects)
+  def self.latest_for(user, count: 5)
+    limit(count)
+      .newest_first
+      .includes(:project, :author)
```
Comment (myabc) [inline app/models/news.rb:76]: Referencing comments from @NobodysNightmare (lost on rebase): https://github.com/myabc/openproject/commit/fa011e178582051403ef60b63195b57d14d759b9#commitcomment-12102033

> I was hoping that these includes statements could also be moved into Project.allowed_to.
> 
> If this is possible: I think you should do it
> If this is not possible: You should at least leave the comment above alive (which is pointing out that mere fact).

My reply (1):

> @NobodysNightmare unfortunately it's not possible to move includes(:project) into Project.allowed_to, as we use Project.allowed_to in places (e.g. Message) where there is no direct belongs_to :project association defined.

My reply (2)

> @NobodysNightmare Not sure I agree about the comment. shouldn't we do that everywhere then?

Comment (NobodysNightmare) [inline app/models/news.rb:76]: The truth is that I hoped you would answer positively on question one ^^

Comment (myabc) [inline app/models/news.rb:76]: @NobodysNightmare not possible :frowning:

File: app/models/project.rb [hunk 2]
```diff
@@ -322,6 +318,17 @@ def self.visible_by(user = nil)
...
 
+  # Returns a ActiveRecord::Relation to find all projects for which +user+ has the given +permission+
+  #
+  # Valid options:
+  # * :project => limit the condition to project
+  # * :with_subprojects => limit the condition to project and its subprojects
+  # * :member => limit the condition to the user projects
+  def self.allowed_to(user, permission, options = {})
+    where(allowed_to_condition(user, permission, options))
+      .references(:projects)
```
Comment (ulferts) [inline app/models/project.rb:329]: I can understand (barely) the need for `references` on the other models when using a string base condition. But that definitely shouldn't be necessary when being in the Project model already.

**Comment by ulferts** [comment]
Behaviour looks good. I'd like to get rid of that `references` and am willing to merge after that.

File: app/models/project.rb (hunk 2 above)
Comment (myabc) [inline app/models/project.rb:329]: @ulferts you're right. it shouldn't be needed - but without it, deprecation warnings are emitted (same problem as https://github.com/opf/openproject/pull/3194#commitcomment-12104671)

**Pull request merged**